felix/fv/bpf_test.go

```diff
@@ -1912,6 +1914,253 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
+						// Use Felix[1] as a middlebox initially.
+						ipRoute := []string{"ip"}
+						if testOpts.ipv6 {
+							ipRoute = append(ipRoute, "-6")
+						}
+
+						cmdCleanRt := append(ipRoute, "route", "del", clusterIP)
+						_ = externalClient.ExecMayFail(strings.Join(cmdCleanRt, ""))
+						cmdCleanRt = append(ipRoute, "route", "del", externalIP)
+						_ = externalClient.ExecMayFail(strings.Join(cmdCleanRt, ""))
+
+						cmdCIP := append(ipRoute, "route", "add", clusterIP, "via", felixIP(1))
+						externalClient.Exec(cmdCIP...)
+						cmdEIP := append(ipRoute, "route", "add", externalIP, "via", felixIP(1))
+						externalClient.Exec(cmdEIP...)
+					})
+
+					It("should have connectivity from external client to maglev backend via cluster IP and external IP", func() {
+						cc.ExpectSome(externalClient, TargetIP(clusterIP), port)
+						cc.ExpectSome(externalClient, TargetIP(externalIP), port)
+						cc.CheckConnectivity()
+					})
+
+					testFailover := func(serviceIP string) {
+						By("making a connection over a loadbalancer and then switching off routing to it")
+						pc := &PersistentConnection{
+							Runtime:             externalClient,
+							RuntimeName:         externalClient.Name,
+							IP:                  serviceIP,
+							Port:                int(port),
+							SourcePort:          50000,
+							Protocol:            testOpts.protocol,
+							MonitorConnectivity: true,
+						}
+						err := pc.Start()
+						Expect(err).NotTo(HaveOccurred())
+						defer pc.Stop()
+
+						Eventually(pc.PongCount, "5s", "100ms").Should(BeNumerically(">", 0), "Connection failed")
+
+						backingPodIPAddr := net.ParseIP(w[0][0].IP)
+						clientIPAddr := net.ParseIP(containerIP(externalClient))
+
+						ctVal, ctExists := checkConntrackExistsAnyDirection(tc.Felixes[1], clientIPAddr, uint16(pc.SourcePort), backingPodIPAddr, uint16(tgtPort), family)
+						Expect(ctExists).To(BeTrue(), "No conntrack (src->dst / dst->src) existed for the connection on Felix[1]")
+						Expect(ctVal.OrigIP().String()).To(Equal(serviceIP), "Unexpected OrigIP on loadbalancer Felix service connection")
+
+						ctVal, ctExists = checkConntrackExistsAnyDirection(tc.Felixes[2], clientIPAddr, uint16(pc.SourcePort), backingPodIPAddr, uint16(tgtPort), family)
+						Expect(ctExists).To(BeFalse(), "Conntrack existed for the connection on Felix[2] before Felix[2] should have handled the connection: %v", ctVal)
+
+						// Traffic is flowing over LB 1. Change ExtClient's serviceIP route to go via LB 2.
+						ipRoute := []string{"ip"}
+						if testOpts.ipv6 {
+							ipRoute = append(ipRoute, "-6")
+						}
+						ipRouteReplace := append(ipRoute, "route", "replace", serviceIP, "via", felixIP(2))
+						externalClient.Exec(ipRouteReplace...)
+
+						lastPongCount := pc.PongCount()
+						Eventually(pc.PongCount, "5s", "100ms").Should(BeNumerically(">", lastPongCount), "Connection is no longer ponging after route failover")
```

Comment:
I think this should get after you check that the connection failed over to make a proof that the connection continues.

Also perhaps there should be a tcpdump probe that makes sure that the connection is not going through the original node. Perhaps watching vxlan packets and making sure their source changes. The conntrack test is a good indication.